File: packages/levelplay/ios/HybridLevelPlayAds.swift
import Foundation
import NitroModules
// Verified by compiling a throwaway Swift file (`swiftc -typecheck`) against the
// real IronSourceSDK 9.5.0.0 XCFramework (https://trunk.cocoapods.org, pod
// "IronSourceSDK") — the framework module is named `IronSource`, and the unified
// LevelPlay iOS API lives under the `LPM*` prefix (LPMInterstitialAd, LPMRewardedAd,
// LPMInitRequestBuilder, LPMPrivacySettings, ...), distinct from the legacy
// `IronSourceAds` facade.
@preconcurrency import IronSource

final class HybridLevelPlayAds: HybridLevelPlayAdsSpec, @unchecked Sendable {
  private let stateLock = NSLock()
  private var isInitialized = false
  private var isInitializing = false
  private var initializedAppKey: String?
  private var initializingAppKey: String?
  private var initializationContinuations: [CheckedContinuation<Void, Error>] = []

  // Keyed by adUnitId: load() and show() are separate calls in this module's
  // spec, so the loaded ad object has to be held onto between the two calls.
  private var loadedInterstitials: [String: InterstitialAdHolder] = [:]
  private var loadedRewarded: [String: RewardedAdHolder] = [:]

  // NSLock.lock()/unlock() are `noasync` in Swift 6 — this keeps locking out of
  // `async` function bodies (a suspension point could otherwise occur while held).
  private func withStateLock<T>(_ body: () -> T) -> T {
    stateLock.lock()
    defer { stateLock.unlock() }
    return body()
  }

  func initialize(appKey: String, testMode: Bool) throws -> Promise<Void> {
    guard !appKey.isEmpty else {
      throw levelPlayAdsError("LevelPlay app key is empty")
    }

    return Promise.async {
      try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.main.async {
          self.stateLock.lock()
          if self.isInitialized {
            let existing = self.initializedAppKey
            self.stateLock.unlock()
            if let existing, existing != appKey {
              continuation.resume(
                throwing: levelPlayAdsError("LevelPlay SDK was already initialized with another app key")
              )
            } else {
              continuation.resume()
            }
            return
          }

          if self.isInitializing, self.initializingAppKey != appKey {
            self.stateLock.unlock()
            continuation.resume(
              throwing: levelPlayAdsError("LevelPlay SDK is being initialized with another app key")
            )
            return
          }

          self.initializationContinuations.append(continuation)
          if self.isInitializing {
            self.stateLock.unlock()
            return
          }
          self.isInitializing = true
          self.initializingAppKey = appKey
          self.stateLock.unlock()

          if testMode {
            LevelPlay.setAdaptersDebug(true)
          }

          let request = LPMInitRequestBuilder(appKey: appKey).build()
          LevelPlay.initWithRequest(request) { _, error in
            self.handleInitialization(error: error)
          }
        }
      }
    }
  }

  fileprivate func handleInitialization(error: Error?) {
    stateLock.lock()
    let continuations = initializationContinuations
    initializationContinuations.removeAll()
    isInitializing = false
    let appKey = initializingAppKey
    initializingAppKey = nil
    if error == nil {
      isInitialized = true
      initializedAppKey = appKey
    }
    stateLock.unlock()

    continuations.forEach { continuation in
      if let error {
        continuation.resume(throwing: error)
      } else {
        continuation.resume()
      }
    }
  }

  func load(adType: LevelPlayAdType, adUnitId: String) throws -> Promise<Void> {
    guard !adUnitId.isEmpty else {
      throw levelPlayAdsError("LevelPlay ad unit ID is empty")
    }

    return Promise.async {
      try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        switch adType {
        case .interstitial:
          DispatchQueue.main.async {
            let holder = InterstitialAdHolder(adUnitId: adUnitId)
            RetainBox.retain(holder)
            holder.onLoaded = { [weak self] in
              self?.withStateLock { self?.loadedInterstitials[adUnitId] = holder }
              RetainBox.release(holder)
              continuation.resume()
            }
            holder.onLoadFailed = { error in
              RetainBox.release(holder)
              continuation.resume(throwing: error)
            }
            holder.ad.loadAd()
          }
        case .rewarded:
          DispatchQueue.main.async {
            let holder = RewardedAdHolder(adUnitId: adUnitId)
            RetainBox.retain(holder)
            holder.onLoaded = { [weak self] in
              self?.withStateLock { self?.loadedRewarded[adUnitId] = holder }
              RetainBox.release(holder)
              continuation.resume()
            }
            holder.onLoadFailed = { error in
              RetainBox.release(holder)
              continuation.resume(throwing: error)
            }
            holder.ad.loadAd()
          }
        }
      }
    }
  }

  func show(adType: LevelPlayAdType, adUnitId: String) throws -> Promise<LevelPlayAdShowResult> {
    guard !adUnitId.isEmpty else {
      throw levelPlayAdsError("LevelPlay ad unit ID is empty")
    }

    return Promise.async {
      let viewController = try await currentViewController()

      switch adType {
      case .interstitial:
        let holder = self.withStateLock { self.loadedInterstitials.removeValue(forKey: adUnitId) }
        guard let holder else {
          throw levelPlayAdsError("LevelPlay interstitial ad is not ready")
        }
        // Reuse the same delegate that was registered in load() — swapping in a
        // different delegate instance here is what used to drop didCloseAd.
        RetainBox.retain(holder)
        return try await withCheckedThrowingContinuation {
          (continuation: CheckedContinuation<LevelPlayAdShowResult, Error>) in
          holder.onDisplayFailed = { error in onShowDisplayFailed(holder, continuation, error) }
          holder.onClosed = {
            RetainBox.release(holder)
            continuation.resume(returning: LevelPlayAdShowResult(state: .completed))
          }
          DispatchQueue.main.async {
            holder.ad.showAd(viewController: viewController, placementName: nil)
          }
        }
      case .rewarded:
        let holder = self.withStateLock { self.loadedRewarded.removeValue(forKey: adUnitId) }
        guard let holder else {
          throw levelPlayAdsError("LevelPlay rewarded ad is not ready")
        }
        RetainBox.retain(holder)
        // Rewarded ads don't carry a Unity-style "finish state" — LevelPlay fires a
        // separate didReward event before didCloseAd, so completion is derived from
        // whether that event happened at all before the ad closed.
        var didReceiveReward = false
        return try await withCheckedThrowingContinuation {
          (continuation: CheckedContinuation<LevelPlayAdShowResult, Error>) in
          holder.onReward = { didReceiveReward = true }
          holder.onDisplayFailed = { error in onShowDisplayFailed(holder, continuation, error) }
          holder.onClosed = {
            RetainBox.release(holder)
            let state: LevelPlayAdShowState = didReceiveReward ? .completed : .skipped
            continuation.resume(returning: LevelPlayAdShowResult(state: state))
          }
          DispatchQueue.main.async {
            holder.ad.showAd(viewController: viewController, placementName: nil)
          }
        }
      }
    }
  }

  func setGDPRConsent(optIn: Bool) throws {
    LPMPrivacySettings.setGDPRConsent(optIn)
  }

  func setCCPAConsent(optIn: Bool) throws {
    // CCPA flag semantics are "opted out of sale", the inverse of our `optIn`
    // (personalized-ads-allowed) flag.
    LPMPrivacySettings.setCCPA(!optIn)
  }

  func setCOPPA(isCoppa: Bool) throws {
    LPMPrivacySettings.setCOPPA(isCoppa)
  }
}

private func onShowDisplayFailed<T>(
  _ holder: AnyObject,
  _ continuation: CheckedContinuation<T, Error>,
  _ error: Error
) {
  RetainBox.release(holder)
  continuation.resume(throwing: error)
}

// LevelPlay requires the same delegate instance to stay registered from
// loadAd() through showAd() — swapping in a different delegate right before
// showAd() (as the old per-phase Load/Show delegate pairs did) is what dropped
// didCloseAd on Android's equivalent path once the listener was swapped.
private class AdHolder: NSObject, @unchecked Sendable {
  var onLoaded: (() -> Void)?
  var onLoadFailed: ((Error) -> Void)?
  var onDisplayed: (() -> Void)?
  var onDisplayFailed: ((Error) -> Void)?
  var onClosed: (() -> Void)?
}

private final class InterstitialAdHolder: AdHolder, LPMInterstitialAdDelegate {
  let ad: LPMInterstitialAd

  init(adUnitId: String) {
    self.ad = LPMInterstitialAd(adUnitId: adUnitId)
    super.init()
    self.ad.setDelegate(self)
  }

  func didLoadAd(with adInfo: LPMAdInfo) { onLoaded?() }
  func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) { onLoadFailed?(error) }
  func didDisplayAd(with adInfo: LPMAdInfo) { onDisplayed?() }
  func didFailToDisplayAd(withAdInfo adInfo: LPMAdInfo, error: Error) { onDisplayFailed?(error) }
  func didClickAd(withAdInfo adInfo: LPMAdInfo) {}
  func didCloseAd(withAdInfo adInfo: LPMAdInfo) { onClosed?() }
}

private final class RewardedAdHolder: AdHolder, LPMRewardedAdDelegate {
  let ad: LPMRewardedAd
  var onReward: (() -> Void)?

  init(adUnitId: String) {
    self.ad = LPMRewardedAd(adUnitId: adUnitId)
    super.init()
    self.ad.setDelegate(self)
  }

  func didLoadAd(with adInfo: LPMAdInfo) { onLoaded?() }
  func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) { onLoadFailed?(error) }
  func didDisplayAd(with adInfo: LPMAdInfo) { onDisplayed?() }
  func didReward(withAdInfo adInfo: LPMAdInfo, reward: LPMReward) { onReward?() }
  func didFailToDisplayAd(withAdInfo adInfo: LPMAdInfo, error: Error) { onDisplayFailed?(error) }
  func didClickAd(withAdInfo adInfo: LPMAdInfo) {}
  func didCloseAd(withAdInfo adInfo: LPMAdInfo) { onClosed?() }
}

File: packages/levelplay/android/src/main/java/com/margelo/nitro/levelplayads/HybridLevelPlayAds.kt
package com.margelo.nitro.levelplayads

import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.unity3d.mediation.LevelPlay
import com.unity3d.mediation.LevelPlayAdError
import com.unity3d.mediation.LevelPlayAdInfo
import com.unity3d.mediation.LevelPlayInitError
import com.unity3d.mediation.LevelPlayInitRequest
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAd
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAdListener
import com.unity3d.mediation.rewarded.LevelPlayReward
import com.unity3d.mediation.rewarded.LevelPlayRewardedAd
import com.unity3d.mediation.rewarded.LevelPlayRewardedAdListener
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

// Verified against the real com.ironsource.sdk:mediationsdk:9.2.0 AAR (javap on
// the downloaded classes.jar, https://android-sdk.is.com/) — the unified LevelPlay
// API lives under `com.unity3d.mediation`. LevelPlayInitRequest.Builder has no
// withLegacyAdFormats() in 9.2.0 (removed since the previous 8.6.0 generation),
// and setConsent/setMetaData/setAdaptersDebug now live on `LevelPlay` itself
// rather than the legacy `com.ironsource.mediationsdk.IronSource` facade, whose
// surface was stripped down in this release.
@DoNotStrip
@Keep
class HybridLevelPlayAds : HybridLevelPlayAdsSpec() {
  private companion object {
    const val TAG = "LevelPlayAds"
  }

  private val context: ReactApplicationContext by lazy {
    NitroModules.applicationContext as? ReactApplicationContext
      ?: error("LevelPlay requires a ReactApplicationContext")
  }

  private val stateLock = Any()
  private var isInitialized = false
  private var initializedAppKey: String? = null
  private var initializingAppKey: String? = null
  private var initialization: CompletableDeferred<Unit>? = null

  // LevelPlay requires the SDK listener to be set exactly once, before the first
  // loadAd() call, and reused for the ad's whole lifetime. Re-registering a
  // different listener object right before showAd() (as this module used to do)
  // silently drops onAdClosed on Android: the internal IronSource event fires
  // (confirmed via logcat, tag "IronSource"), but the unified
  // LevelPlayInterstitialAdListener/LevelPlayRewardedAdListener never sees it
  // once its listener has been swapped. So the SDK listener is installed once
  // in the holder's init{} block, and load()/show() only swap out our own
  // plain-Kotlin callback fields, never the SDK listener.
  private abstract class AdHolder {
    var onLoaded: (() -> Unit)? = null
    var onLoadFailed: ((LevelPlayAdError) -> Unit)? = null
    var onDisplayed: (() -> Unit)? = null
    var onDisplayFailed: ((LevelPlayAdError) -> Unit)? = null
    var onClosed: (() -> Unit)? = null
  }

  private class InterstitialHolder(adUnitId: String) : AdHolder() {
    val ad = LevelPlayInterstitialAd(adUnitId)

    init {
      ad.setListener(object : LevelPlayInterstitialAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) = onLoaded?.invoke() ?: Unit
        override fun onAdLoadFailed(error: LevelPlayAdError) = onLoadFailed?.invoke(error) ?: Unit
        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) = onDisplayed?.invoke() ?: Unit
        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) =
          onDisplayFailed?.invoke(error) ?: Unit
        override fun onAdClosed(adInfo: LevelPlayAdInfo) = onClosed?.invoke() ?: Unit
      })
    }
  }

  private class RewardedHolder(adUnitId: String) : AdHolder() {
    val ad = LevelPlayRewardedAd(adUnitId)
    var onRewarded: (() -> Unit)? = null

    init {
      ad.setListener(object : LevelPlayRewardedAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) = onLoaded?.invoke() ?: Unit
        override fun onAdLoadFailed(error: LevelPlayAdError) = onLoadFailed?.invoke(error) ?: Unit
        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) = onDisplayed?.invoke() ?: Unit
        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) =
          onDisplayFailed?.invoke(error) ?: Unit
        override fun onAdRewarded(reward: LevelPlayReward, adInfo: LevelPlayAdInfo) = onRewarded?.invoke() ?: Unit
        override fun onAdClosed(adInfo: LevelPlayAdInfo) = onClosed?.invoke() ?: Unit
      })
    }
  }

  // Keyed by adUnitId: load() and show() are separate calls in this module's
  // spec, so the loaded ad object has to be held onto between the two calls.
  private val loadedInterstitials = mutableMapOf<String, InterstitialHolder>()
  private val loadedRewarded = mutableMapOf<String, RewardedHolder>()

  override fun initialize(appKey: String, testMode: Boolean): Promise<Unit> = Promise.async {
    require(appKey.isNotBlank()) { "LevelPlay app key is empty" }

    val (deferred, shouldStart) = synchronized(stateLock) {
      if (isInitialized) {
        require(initializedAppKey == appKey) {
          "LevelPlay SDK was already initialized with another app key"
        }
        return@async
      }

      val current = initialization
      if (current != null) {
        require(initializingAppKey == appKey) {
          "LevelPlay SDK is being initialized with another app key"
        }
        current to false
      } else {
        CompletableDeferred<Unit>().also {
          initialization = it
          initializingAppKey = appKey
        } to true
      }
    }

    if (shouldStart) {
      withContext(Dispatchers.Main) {
        if (testMode) {
          LevelPlay.setAdaptersDebug(true)
        }

        val initRequest = LevelPlayInitRequest.Builder(appKey).build()

        LevelPlay.init(
          context,
          initRequest,
          object : com.unity3d.mediation.LevelPlayInitListener {
            override fun onInitSuccess(configuration: com.unity3d.mediation.LevelPlayConfiguration) {
              synchronized(stateLock) {
                isInitialized = true
                initializedAppKey = appKey
                initialization = null
                initializingAppKey = null
              }
              deferred.complete(Unit)
            }

            override fun onInitFailed(error: LevelPlayInitError) {
              synchronized(stateLock) {
                initialization = null
                initializingAppKey = null
              }
              deferred.completeExceptionally(
                levelPlayAdsException("initialize", error.errorCode.toString(), error.errorMessage),
              )
            }
          },
        )
      }
    }

    deferred.await()
  }

  override fun load(adType: LevelPlayAdType, adUnitId: String): Promise<Unit> = Promise.async {
    require(adUnitId.isNotBlank()) { "LevelPlay ad unit ID is empty" }
    val deferred = CompletableDeferred<Unit>()

    withContext(Dispatchers.Main) {
      when (adType) {
        LevelPlayAdType.INTERSTITIAL -> {
          val holder = InterstitialHolder(adUnitId)
          holder.onLoaded = {
            synchronized(stateLock) { loadedInterstitials[adUnitId] = holder }
            deferred.complete(Unit)
          }
          holder.onLoadFailed = { error ->
            deferred.completeExceptionally(
              levelPlayAdsException("load", error.getErrorCode().toString(), error.getErrorMessage()),
            )
          }
          holder.ad.loadAd()
        }
        LevelPlayAdType.REWARDED -> {
          val holder = RewardedHolder(adUnitId)
          holder.onLoaded = {
            synchronized(stateLock) { loadedRewarded[adUnitId] = holder }
            deferred.complete(Unit)
          }
          holder.onLoadFailed = { error ->
            deferred.completeExceptionally(
              levelPlayAdsException("load", error.getErrorCode().toString(), error.getErrorMessage()),
            )
          }
          holder.ad.loadAd()
        }
      }
    }

    deferred.await()
  }

  private fun onShowDisplayFailed(
    kind: String,
    adUnitId: String,
    error: LevelPlayAdError,
    deferred: CompletableDeferred<*>,
  ) {
    Log.d(TAG, "show($kind, $adUnitId): onAdDisplayFailed (${error.getErrorCode()}) ${error.getErrorMessage()}")
    deferred.completeExceptionally(
      levelPlayAdsException("show", error.getErrorCode().toString(), error.getErrorMessage()),
    )
  }

  override fun show(adType: LevelPlayAdType, adUnitId: String): Promise<LevelPlayAdShowResult> = Promise.async {
    require(adUnitId.isNotBlank()) { "LevelPlay ad unit ID is empty" }
    val activity = context.requireCurrentActivity()
    val deferred = CompletableDeferred<LevelPlayAdShowResult>()

    when (adType) {
      LevelPlayAdType.INTERSTITIAL -> {
        val holder = synchronized(stateLock) { loadedInterstitials.remove(adUnitId) }
          ?: error("LevelPlay interstitial ad is not ready")
        // Reuse the same SDK listener that was registered in load() — swapping
        // in a different listener object here is what used to drop onAdClosed.
        holder.onLoaded = null
        holder.onLoadFailed = null
        holder.onDisplayed = {
          Log.d(TAG, "show(interstitial, $adUnitId): onAdDisplayed")
        }
        holder.onDisplayFailed = { error -> onShowDisplayFailed("interstitial", adUnitId, error, deferred) }
        holder.onClosed = {
          Log.d(TAG, "show(interstitial, $adUnitId): onAdClosed")
          deferred.complete(LevelPlayAdShowResult(LevelPlayAdShowState.COMPLETED))
        }
        Log.d(TAG, "show(interstitial, $adUnitId): calling showAd")
        withContext(Dispatchers.Main) { holder.ad.showAd(activity) }
      }
      LevelPlayAdType.REWARDED -> {
        val holder = synchronized(stateLock) { loadedRewarded.remove(adUnitId) }
          ?: error("LevelPlay rewarded ad is not ready")
        // Rewarded ads don't carry a Unity-style "finish state" — LevelPlay fires a
        // separate onAdRewarded event before onAdClosed, so completion is derived
        // from whether that event happened at all before the ad closed.
        var didReceiveReward = false
        holder.onLoaded = null
        holder.onLoadFailed = null
        holder.onDisplayed = {
          Log.d(TAG, "show(rewarded, $adUnitId): onAdDisplayed")
        }
        holder.onRewarded = {
          Log.d(TAG, "show(rewarded, $adUnitId): onAdRewarded")
          didReceiveReward = true
        }
        holder.onDisplayFailed = { error -> onShowDisplayFailed("rewarded", adUnitId, error, deferred) }
        holder.onClosed = {
          Log.d(TAG, "show(rewarded, $adUnitId): onAdClosed (didReceiveReward=$didReceiveReward)")
          val state = if (didReceiveReward) {
            LevelPlayAdShowState.COMPLETED
          } else {
            LevelPlayAdShowState.SKIPPED
          }
          deferred.complete(LevelPlayAdShowResult(state))
        }
        Log.d(TAG, "show(rewarded, $adUnitId): calling showAd")
        withContext(Dispatchers.Main) { holder.ad.showAd(activity) }
      }
    }

    deferred.await()
  }

  override fun setGDPRConsent(optIn: Boolean) {
    LevelPlay.setConsent(optIn)
  }

  // CCPA flag semantics are "opted out of sale", the inverse of our `optIn`
  // (personalized-ads-allowed) flag.
  override fun setCCPAConsent(optIn: Boolean) {
    LevelPlay.setMetaData("do_not_sell", if (optIn) "false" else "true")
  }

  override fun setCOPPA(isCoppa: Boolean) {
    LevelPlay.setMetaData("is_child_directed", if (isCoppa) "true" else "false")
  }
}
